### ISSUE TITLE
ORCH/SR cannot find service due to services with different metadata

### ISSUE THREAD

**Issue by eudyptula** [arrowhead-f/core-java#49]
It all started by ORCH/SR not finding the registered service (in case others have a similar problem). To check if this really is the issue, you can try disabling metadata search in the consumer, that should avoid the buggy code.

Now, the real issue is that two Arrowhead services (of the same service definition) cannot differ in metadata - or at the very least it corrupts the search that the ORCH/SR is doing to find them. Looking at the database, metadata is on the service registry entries, not the Arrowhead service entries. I will also argue that this is probably correct.

However, in the code it is quite common to find the metadata field in the service classes, meaning that in the communication protocol  the metadata is attached to service entries and thus can no longer differ between services of the same definition. Thinking that the solution is to move metadata from the service class to the service registry entry and to the service request form (so you can still search for services with specific metadata.

Just as a technical note on the original issue, the real cause of the failed search is the [ServiceRegistryEntry::fromDatabase()](https://github.com/arrowhead-f/core-java/blob/develop/core-common/src/main/java/eu/arrowhead/common/database/ServiceRegistryEntry.java#L196) method. Doing a service registry search multiple entries may point to the same provided service, and this method will blinded override the metadata in this service. Causing a mismatch is metadata and the search to fail.

Now, I wanted to create a fix for this, but realised that it is not that simple:
1. Communication protocol in Arrowhead will change as I need to move the metadata field from the service to the request form and service entry.
2. The code is currently duplicated in both the core and client repositories (and across multiple branches), and the code has been changed in some of them. We should really work on merging this code, would make it a lot easier to fix things like this.

**Comment by hegeduscs** [comment]
The DTOs and DSOs are different in this case, yes. In the REST interface, service metadata is inside the service object, while in storage it belongs to the SREntry purposefully. 

I do not promote going way #1, because it was intentional, and makes more sense to have _service_ metadata in the service object. However, storage is different in this case, that given metadata belongs to a service provider _instance_ and cannot be stored generically within the arrowhead_service table to avoid overwrites. 

Metadata handling is PoC level, hence the flag to turn it off during orchestration. It was intended in my design to support semantic web stlye thinking.

**Comment by eudyptula** [comment]
Maybe we need to distinguish between _AbstractService_ and _ConcreteService_ then. At least in the service registry / database level. Let me look at little more into this...

**Comment by hegeduscs** [comment]
No, we need a Data Transfer Object Model and a Data Storage Object Model notation, as per the generic Spring development terminology. Do not start using AbstractService and ConcreteService terms, because these will be confusing, for everyone. A service is an abstraction on its own, not an endpoint, in general.
